lib/tests/formulae.rb

```diff
@@ -607,51 +611,49 @@ def formula!(formula_name, args:)
         test "brew", "install", "--only-dependencies", *install_args,
               env:  { "HOMEBREW_DEVELOPER" => nil },
               args: args
+
         test "brew", "install", *install_args,
               env:  { "HOMEBREW_DEVELOPER" => nil },
               args: args
         install_passed = steps.last.passed?
 
         test "brew", "audit", *audit_args, args: args
+        return unless install_passed
 
-        if install_passed
-          bottle_reinstall_formula(formula, new_formula, args: args)
-          test "brew", "linkage", "--test", formula_name, args: args
-
-          test_passed_or_missing = if formula.test_defined?
-            test "brew", "install", "--only-dependencies", "--include-test",
-                                    formula_name,
-                                    args: args
-            # Intentionally not passing --retry here to avoid papering over
-            # flaky tests when a formula isn't being pulled in as a dependent.
-            test "brew", "test", "--verbose", formula_name, args: args
-
-            steps.last.passed?
-          else
-            true
-          end
+        bottle_reinstall_formula(formula, new_formula, args: args)
+        test "brew", "linkage", "--test", formula_name, args: args
+        failed_linkage_or_test = steps.last.failed?
 
-          # Don't test dependents if the formula test failed.
-          if test_passed_or_missing
-            @source_dependents.each do |dependent|
-              install_dependent_from_source(dependent, args: args)
+        test "brew", "install", "--only-dependencies", "--include-test",
+                                formula_name,
+                                args: args
+        # Intentionally not passing --retry here to avoid papering over
+        # flaky tests when a formula isn't being pulled in as a dependent.
+        test "brew", "test", "--verbose", formula_name, args: args
+        failed_linkage_or_test ||= steps.last.failed?
```

Comment:
What's the behaviour of this if a formula doesn't have a test defined? Previously this had a `formula.test_defined?` check.

I know this should ideally not be the case, but it isn't something we yet mandate from existing formulae yet (still a `--strict`-only audit).

Reply:
Re-added check.